File: scripts/ws-test/test-terminal.py
#!/usr/bin/env python3
"""
Terminal WebSocket Test Script (Python)

Simple Python script to test the terminal WebSocket API without
wrestling with bash escaping issues.

Usage:
    python3 test-terminal.py "pwd"
    python3 test-terminal.py "echo Hello World"
    python3 test-terminal.py "ls -la"

Install dependencies:
    pip3 install websocket-client
"""

import sys
import json
import time
import signal
from websocket import create_connection, WebSocketTimeoutException

# Configuration
WS_URL = "ws://localhost:8080/api/v1/terminal/session"
TIMEOUT_SECONDS = 5

def send_command(command):
    """Send a command to the terminal WebSocket API."""
    
    print(f"\n🔌 Connecting to {WS_URL}...")
    print(f"📝 Command: {command}\n")
    
    try:
        # Create WebSocket connection with timeout
        ws = create_connection(WS_URL, timeout=TIMEOUT_SECONDS)
        print("✅ Connected!\n")
        print("📤 Sending command...\n")
        
        # Send command with CR+LF
        message = {
            "type": "input",
            "data": command + "\r\n"
        }
        ws.send(json.dumps(message))
        
        # Receive and print output
        start_time = time.time()
        while True:
            # Check timeout
            if time.time() - start_time > TIMEOUT_SECONDS:
                print("\n⏱️  Timeout reached, closing connection...\n")
                break
            
            try:
                # Receive message with short timeout for responsiveness
                result = ws.recv()
                message = json.loads(result)
                
                if message["type"] == "output":
                    # Print output data (may contain ANSI codes)
                    print(message["data"], end="", flush=True)
                    
                elif message["type"] == "error":
                    print(f"\n❌ Error: {message['message']}\n", file=sys.stderr)
                    
                elif message["type"] == "exit":
                    print(f"\n🏁 Process exited with code: {message.get('code', 0)}\n")
                    break
                    
                else:
                    print(f"\n⚠️  Unknown message type: {message['type']}")
                    
            except WebSocketTimeoutException:
                # No message available, continue
                continue
            except json.JSONDecodeError as e:
                print(f"\n❌ Failed to parse message: {e}\n", file=sys.stderr)
                print(f"Raw data: {result}\n", file=sys.stderr)
                
        ws.close()
        print("\n🔌 Connection closed\n")
        
    except Exception as e:
        print(f"\n❌ Error: {e}\n", file=sys.stderr)
        sys.exit(1)

def signal_handler(sig, frame):
    """Handle Ctrl+C gracefully."""
    print("\n\n⚠️  Interrupted, exiting...\n")
    sys.exit(0)

if __name__ == "__main__":
    # Register Ctrl+C handler
    signal.signal(signal.SIGINT, signal_handler)
    
    # Get command from command line
    if len(sys.argv) < 2:
        command = "pwd"
        print("ℹ️  No command provided, using default: pwd")
    else:
        command = sys.argv[1]
    
    send_command(command)

File: scripts/ws-test/test-terminal.js
#!/usr/bin/env node
/**
 * Terminal WebSocket Test Script
 * 
 * Simple Node.js script to test the terminal WebSocket API without
 * wrestling with bash escaping issues.
 * 
 * Usage:
 *   node test-terminal.js "pwd"
 *   node test-terminal.js "echo Hello World"
 *   node test-terminal.js "ls -la"
 * 
 * Install dependencies:
 *   npm install ws
 */

const WebSocket = require('ws');

// Configuration
const WS_URL = 'ws://localhost:8080/api/v1/terminal/session';
const TIMEOUT_MS = 5000;

// Get command from command line
const command = process.argv[2] || 'pwd';

console.log(`\n🔌 Connecting to ${WS_URL}...`);
console.log(`📝 Command: ${command}\n`);

// Create WebSocket connection
const ws = new WebSocket(WS_URL);

// Track if we've sent the command
let commandSent = false;

// Timeout to close connection
const timeout = setTimeout(() => {
  console.log('\n⏱️  Timeout reached, closing connection...\n');
  ws.close();
}, TIMEOUT_MS);

ws.on('open', () => {
  console.log('✅ Connected!\n');
  console.log('📤 Sending command...\n');
  
  // Send command with CR+LF
  const message = {
    type: 'input',
    data: command + '\r\n'
  };
  
  ws.send(JSON.stringify(message));
  commandSent = true;
});

ws.on('message', (data) => {
  try {
    const message = JSON.parse(data.toString());
    
    switch (message.type) {
      case 'output':
        // Print output data (may contain ANSI codes)
        process.stdout.write(message.data);
        break;
        
      case 'error':
        console.error(`\n❌ Error: ${message.message}\n`);
        break;
        
      case 'exit':
        console.log(`\n🏁 Process exited with code: ${message.code}\n`);
        ws.close();
        break;
        
      default:
        console.log(`\n⚠️  Unknown message type: ${message.type}`);
    }
  } catch (err) {
    console.error(`\n❌ Failed to parse message: ${err.message}`);
    console.error(`Raw data: ${data.toString()}\n`);
  }
});

ws.on('error', (err) => {
  console.error(`\n❌ WebSocket error: ${err.message}\n`);
  process.exit(1);
});

ws.on('close', () => {
  clearTimeout(timeout);
  console.log('\n🔌 Connection closed\n');
  process.exit(0);
});

// Handle Ctrl+C
process.on('SIGINT', () => {
  console.log('\n\n⚠️  Interrupted, closing connection...\n');
  ws.close();
});
